modules/social_features/social_core/src/Service/NodeModerationMenuHelperService.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Drupal\social_core\Service;
+
+/**
+ * Class NodeModerationMenuHelperService.
+ *
+ * @package Drupal\social_core\Service
+ */
+class NodeModerationMenuHelperService {
+
+  /**
+   * Add a new link to existing moderation node menu.
+   *
+   * @param array $existing_links
+   *   The moderation node menu.
+   * @param string $new_link_name
+   *   The new link name.
+   * @param array $new_link
+   *   The new link we.
+   *
+   * @return void
+   *   Changed menu array.
+   */
+  public function addNodeModerationLink(array &$existing_links, string $new_link_name, array $new_link,): void {
+    // Implemented this array in a such way because $links['moderation'] array
+    // can be used in many places.
+    if (!isset($existing_links['moderation'])) {
+      $existing_links['moderation'] = [
+        '#theme' => 'links__node__moderation',
+        '#attributes' => [
+          'class' => [
+            'links',
+            'inline',
+          ],
+        ],
+      ];
+    }
+
+    $existing_links['moderation']['#links'][$new_link_name] = $new_link;
+  }
+
+}
```

Comment:
If we're introducing services like this then I think we should provide some motivation for how we expect the service to behave and when developers should use it. I think in general anything named "HelperService" is a code-smell because we apparently don't know how to give it a focused enough name and are just trying to de-duplicate code.

Specifically for this service I don't think it should exist:

- The amount of code we're replacing is very small and we're creating an abstraction around Drupal render arrays, which itself is already an abstraction layer.
- It's modifying an array that's passed in while code that works with render arrays may want to edit it. That means the calling code needs to know exactly what happens inside the method, so it's a leaky abstraction.

It took me until here to understand that what we're doing here is not actually adding to the `node_moderation` menu, but we're changing the purpose of the `moderation` menu to be a generic "contextual actions" menu. That's also why you're trying to add something specific to `social_content_report` in `social_core`. 

That is a very strong indication that we should probably re-evaluate what we want and make proper use of Drupal to achieve it (e.g. something like Drupal's `tasks` menu system, or a parallel menu that we create/manage ourselves).